src/_imaging.c

```diff
@@ -3748,6 +3748,138 @@ static PySequenceMethods image_as_sequence = {
+static int
+_compare_pixels(
+    const char *mode, const int ysize, const int linesize,
+    const UINT8 **pixels_a, const UINT8 **pixels_b
+) {
+    // Fortunately, all of the modes that have extra bytes in their pixels use four bytes for their pixels.
+    UINT32 mask = 0xffffffff;
+    if (
+        !strcmp(mode, "RGB") || !strcmp(mode, "YCbCr") ||
+        !strcmp(mode, "HSV") || !strcmp(mode, "LAB")
+    ) {
+        // These modes have three channels in four bytes,
+        // so we have to ignore the last byte.
+#ifdef WORDS_BIGENDIAN
+        mask = 0xffffff00;
+#else
+        mask = 0x00ffffff;
+#endif
+    } else if (!strcmp(mode, "LA") || !strcmp(mode, "La") || !strcmp(mode, "PA")) {
+        // These modes have two channels in four bytes,
+        // so we have to ignore the middle two bytes.
+        mask = 0xff0000ff;
+    }
+
+    if (mask == 0xffffffff) {
+        // If we aren't masking anything we can use memcmp.
+        int y;
+        for (y = 0; y < ysize; y++) {
+            if (memcmp(pixels_a[y], pixels_b[y], linesize)) {
+                return 1;
+            }
+        }
+    } else {
+        const int xsize = linesize / 4;
+        int y, x;
+        for (y = 0; y < ysize; y++) {
+            UINT32 *line_a = (UINT32*)pixels_a[y];
+            UINT32 *line_b = (UINT32*)pixels_b[y];
+            for (x = 0; x < xsize; x++, line_a++, line_b++) {
+                if ((*line_a & mask) != (*line_b & mask)) {
+                    return 1;
+                }
+            }
+        }
+    }
+    return 0;
+}
+
+static PyObject *
+image_richcompare(const ImagingObject *self, const PyObject *other, const int op) {
+    if (op != Py_EQ && op != Py_NE) {
+        Py_RETURN_NOTIMPLEMENTED;
+    }
+
+    // If the other object is not an ImagingObject.
+    if (!PyImaging_Check(other)) {
+        if (op == Py_EQ) {
+            Py_RETURN_FALSE;
+        } else {
+            Py_RETURN_TRUE;
+        }
+    }
+
+    const Imaging img_a = self->image;
+    const Imaging img_b = ((ImagingObject*)other)->image;
+
+    if (
+        strcmp(img_a->mode, img_b->mode)
+        || img_a->xsize != img_b->xsize
+        || img_a->ysize != img_b->ysize
+    ) {
+        if (op == Py_EQ) {
+            Py_RETURN_FALSE;
+        } else {
+            Py_RETURN_TRUE;
+        }
+    }
+
+    const ImagingPalette palette_a = img_a->palette;
+    const ImagingPalette palette_b = img_b->palette;
+    if (palette_a || palette_b) {
+        const UINT8 *palette_a_data = palette_a->palette;
+        const UINT8 *palette_b_data = palette_b->palette;
+        const UINT8 **palette_a_data_ptr = &palette_a_data;
+        const UINT8 **palette_b_data_ptr = &palette_b_data;
+        if (
+            !palette_a || !palette_b
+            || palette_a->size != palette_b->size
+            || strcmp(palette_a->mode, palette_b->mode)
+            || _compare_pixels(
+                palette_a->mode,
+                1,
+                palette_a->size * 4,
```

Comment:
Why is this multiplied by 4?

Reply:
The third parameter is the line size, but palettes don't have a line size, so it has to be calculated from the number of pixels and the pixel size. Palettes also don't have a stored pixel size, but currently it's always 4. It's not good to hardcode it like this, but there isn't anywhere to actually get it from.